File: frontend/src/api/save.ts
import apiClient from './client'
import type { GameState } from '../types/game'

export interface SaveResponse {
  state: GameState
  state_version: number
}

export interface PutSaveRequest {
  state: GameState
  state_version: number
}

export const saveApi = {
  getSave: async (): Promise<SaveResponse> => {
    const response = await apiClient.get('/api/save')
    return response.data
  },

  putSave: async (data: PutSaveRequest): Promise<{ state_version: number }> => {
    const response = await apiClient.put('/api/save', data)
    return response.data
  }
}

File: backend/app/api/save.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.services import save_service
from app.schemas.game_state import GameState

router = APIRouter(tags=["save"])


@router.get("/api/save")
async def get_save(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    result = save_service.get_save(db, current_user.id)
    return result


@router.put("/api/save")
async def put_save(
    data: dict,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    state = data.get("state")
    state_version = data.get("state_version", 0)
    
    result = save_service.put_save(db, current_user.id, state, state_version)
    return result
